File: server/src/main/java/gg/mineral/server/network/packet/play/clientbound/SpawnMobPacket.java
package gg.mineral.server.network.packet.play.clientbound;

import java.util.List;

import gg.mineral.api.network.packet.Packet;
import gg.mineral.server.entity.meta.EntityMetadataImpl;
import io.netty.buffer.ByteBuf;

public final record SpawnMobPacket(
        int entityId,
        short type,
        int x,
        int y,
        int z,
        byte yaw,
        byte pitch,
        byte headPitch,
        short velocityX,
        short velocityY,
        short velocityZ,
        List<EntityMetadataImpl.Entry> entries) implements Packet.OUTGOING {
    @Override
    public void serialize(ByteBuf os) {
        writeVarInt(os, entityId);
        os.writeByte(type);
        os.writeInt(x);
        os.writeInt(y);
        os.writeInt(z);
        os.writeByte(yaw);
        os.writeByte(pitch);
        os.writeByte(headPitch);
        os.writeShort(velocityX);
        os.writeShort(velocityY);
        os.writeShort(velocityZ);
        writeMetadata(os, entries);
    }

    @Override
    public byte getId() {
        return 0x0F;
    }
}

File: server/src/main/kotlin/gg/mineral/server/network/packet/play/clientbound/SpawnMobPacket.kt
package gg.mineral.server.network.packet.play.clientbound

import gg.mineral.api.entity.meta.EntityMetadata
import gg.mineral.api.network.packet.Packet
import io.netty.buffer.ByteBuf

@JvmRecord
data class SpawnMobPacket(
    val entityId: Int,
    val type: Short,
    val x: Int,
    val y: Int,
    val z: Int,
    val yaw: Byte,
    val pitch: Byte,
    val headPitch: Byte,
    val velocityX: Short,
    val velocityY: Short,
    val velocityZ: Short,
    val entries: List<EntityMetadata.Entry>
) : Packet.Outgoing {
    override fun serialize(os: ByteBuf) {
        os.writeVarInt(entityId)
        os.writeByte(type.toInt())
        os.writeInt(x, y, z)
        os.writeByte(yaw, pitch, headPitch)
        os.writeShort(velocityX, velocityY, velocityZ)
        os.writeMetadata(entries)
    }

    override val id: Byte
        get() = 0x0F
}
